pkg/imagestream/imagestream.go

```diff
@@ -0,0 +1,180 @@
+type Builder struct {
+	// imageStream definition, used to create the imageStream object.
+	Definition *imagev1.ImageStream
+	// Created imageStream object.
+	Object *imagev1.ImageStream
+	// api client to interact with the cluster.
+	apiClient goclient.Client
+	// Used in functions that define or mutate clusterOperator definition. errorMsg is processed before the
+	// ClusterOperator object is created.
+	errorMsg string
+}
+
+// Pull retrieves an existing imageStream object from the cluster.
+func Pull(apiClient *clients.Settings, name, nsname string) (*Builder, error) {
+	glog.V(100).Infof(
+		"Pulling imageStream object name %s from namespace %s", name, nsname)
+
+	if apiClient == nil {
+		glog.V(100).Infof("The apiClient is empty")
+
+		return nil, fmt.Errorf("imageStream 'apiClient' cannot be empty")
+	}
+
+	builder := Builder{
+		apiClient: apiClient.Client,
+		Definition: &imagev1.ImageStream{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			},
+			Spec: imagev1.ImageStreamSpec{},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the imageStream is empty")
+
+		return nil, fmt.Errorf("imageStream 'name' cannot be empty")
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the imageStream is empty")
+
+		return nil, fmt.Errorf("imageStream 'nsname' cannot be empty")
+	}
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("imageStream object %s does not exist in namespace %s",
+			name, nsname)
+	}
+
+	builder.Definition = builder.Object
+
+	return &builder, nil
+}
+
+// Get fetches existing imageStream from cluster.
+func (builder *Builder) Get() (*imagev1.ImageStream, error) {
+	if valid, err := builder.validate(); !valid {
+		return nil, err
+	}
+
+	glog.V(100).Infof("Getting existing imageStream with name %s in namespace %s from cluster",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	imageStreamObj := &imagev1.ImageStream{}
+	err := builder.apiClient.Get(context.TODO(), goclient.ObjectKey{
+		Name:      builder.Definition.Name,
+		Namespace: builder.Definition.Namespace,
+	}, imageStreamObj)
+
+	if err != nil {
+		glog.V(100).Infof("imageStream object %s does not exist in namespace %s",
+			builder.Definition.Name, builder.Definition.Namespace)
+
+		return nil, err
+	}
+
+	return imageStreamObj, nil
+}
+
+// Exists checks whether the given imageStream exists.
+func (builder *Builder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if imageStream %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// GetDockerImage fetches imageStream DockerImage value.
+func (builder *Builder) GetDockerImage(imageTag string) (string, error) {
+	if valid, err := builder.validate(); !valid {
+		return "", err
+	}
+
```

Comment:
Can you verify that imageTag is not empty

Reply:
good catch, thank you! missed it; it fixed now